test/MetaGuard.spec.ts

```diff
@@ -0,0 +1,260 @@
+import { AddressZero } from "@ethersproject/constants";
+import { AddressOne } from "@gnosis.pm/safe-contracts";
+import { expect } from "chai";
+import hre, { deployments, waffle, ethers } from "hardhat";
+
+import "@nomiclabs/hardhat-ethers";
+
+describe("MetaGuard", async () => {
+  const [user1, user2] = waffle.provider.getWallets();
+  const abiCoder = new ethers.utils.AbiCoder();
+
+  const setupTests = deployments.createFixture(async ({ deployments }) => {
+    await deployments.fixture();
+    const avatarFactory = await hre.ethers.getContractFactory("TestAvatar");
+    const avatar = await avatarFactory.deploy();
+    const testGuardFactory = await hre.ethers.getContractFactory("TestGuard");
+    const testGuard = await testGuardFactory.deploy(
+      user1.address,
+      avatar.address
+    );
+    const metaGuardFactory = await hre.ethers.getContractFactory("MetaGuard");
+    const metaGuard = await metaGuardFactory.deploy(
+      user1.address,
+      avatar.address,
+      [testGuard.address]
+    );
+    const setGuard = await avatar.populateTransaction.setGuard(
+      metaGuard.address
+    );
+    await expect(
+      avatar.execTransactionFromModule(avatar.address, 0, setGuard.data, 0)
+    );
+    const tx = {
+      to: avatar.address,
+      value: 0,
+      data: "0x",
+      operation: 0,
+      avatarTxGas: 0,
+      baseGas: 0,
+      gasPrice: 0,
+      gasToken: AddressZero,
+      refundReceiver: AddressZero,
+      signatures: "0x",
+    };
+
+    return {
+      avatar,
+      metaGuard,
+      tx,
+      testGuard,
+    };
+  });
+
+  describe("setUp()", async () => {
+    it("throws if guard has already been initialized", async () => {
+      const { avatar, metaGuard } = await setupTests();
+      const initializeParams = abiCoder.encode(
+        ["address", "address", "address[]"],
+        [user1.address, avatar.address, []]
+      );
+      await expect(metaGuard.setUp(initializeParams)).to.be.revertedWith(
+        "Initializable: contract is already initialized"
+      );
+    });
+
+    it("throws if owner is zero address", async () => {
+      const MetaGuard = await hre.ethers.getContractFactory("MetaGuard");
+      await expect(
+        MetaGuard.deploy(AddressZero, AddressZero, [])
+      ).to.be.revertedWith("Ownable: new owner is the zero address");
+    });
+
+    it("should emit event because of successful set up", async () => {
+      const MetaGuard = await hre.ethers.getContractFactory("MetaGuard");
+      const metaGuard = await MetaGuard.deploy(
+        user1.address,
+        user1.address,
+        []
+      );
+      await metaGuard.deployed();
+
+      await expect(metaGuard.deployTransaction)
+        .to.emit(metaGuard, "MetaGuardSetup")
+        .withArgs(user1.address, user1.address, user1.address, []);
+    });
+  });
+
+  describe("fallback", async () => {
+    it("must NOT revert on fallback without value", async () => {
+      const { metaGuard } = await setupTests();
```

Comment:
Maybe `beforeEach` would be useful to clean up the `setupTests()` calls

Reply:
using setupTests instead of beforeEach because I need to use the deployments.fixture functions to make the tests run faster